rfcs/assert_throws.md

```diff
@@ -0,0 +1,45 @@
+# `testharness.js`: Replace `assert_throws` and `promise_rejects` with more fine-grained APIs
+
+## Summary
+
+The expected value passed to `assert_throws` and `promise_rejects` APIs can be one of several different kinds of values.
+This makes it harder to correctly compare the actual thrown exception / rejection value.
+Remove these APIs, and replace them with more focused APIs.
+
+## Details
+
+The `assert_throws` and `promise_rejects` APIs verify the value that is thrown or passed as a rejection value.
+They support being called with three kinds of expected values:
+
+- a string or number representing a `DOMException` (e.g., `"INVALID_STATE_ERR"`, `"InvalidStateError"`, or `DOMException.INVALID_STATE_ERR`);
+- an ECMAScript-style `Error` object (e.g., `new TypeError()`);
+- a plain object (in practice, the exact same object should be thrown or used as the rejection value).
+
+The last two are handled the same way: only the "name" properties are compared between the expected and the actual value.
+Because of the conflation of the various cases, the existing APIs don't get any of them quite right, and this causes some browser bugs to be missed.
+For example, the following passes:
+
+```js
+assert_throws(new TypeError(), () => { throw new DOMException("TypeError") });
+```
+
+Six new APIs are proposed to handle each of the cases separately:
+
+- `promise_rejects_dom` and `assert_throws_dom`, called with a string representing a `DOMException` (as above);
+- `promise_rejects_js` and `assert_throws_js`, called with the Error constructor (e.g. `TypeError`);
+- `assert_throws_exactly` and `promise_rejects_exactly`, called with the expected object and using `Object.is`.
```

Comment:
"expected value".  It doesn't have to be an object.

Reply:
`Object.is` caught my eye here. I don't think we use this elsewhere and yet the benefits would be the same here and elsewhere. I don't think -0 and NaN are more common values to throw, so I wonder if we should use `===` and add in support of `Object.is` to various asserts separately?

Reply:
I'd be fine with using `===`.

Reply:
Or, actually, `assert_equals` for consistency?

Reply:
Looks like `assert_class_string` uses it internally, but most asserts seemingly don't reuse other asserts like this. I think it's because the resulting error message then isn't clear, so this should be avoided?

Reply:
You can catch the error and rethrow with whatever message you like. That's a bit more verbose of course, but `assert_equals` does take some care around edge cases.

Reply:
I sent https://github.com/web-platform-tests/wpt/pull/19219 using `same_value` for `assert_class_string`, see if you like that?

If so I guess let's just use `same_value` here too?

Reply:
We can do that, but `same_value` just implements the `Object.is` semantics (but presumably predates it), right?  So maybe we should consider removing `same_value` altogether and using `Object.is` consistently.  In any case, I'm happy to do whatever here, but I'd really like to get this landed ASAP; at this point I have quite a large number of other changes already written and reviewed but blocked on this...

Reply:
`same_value` is indeed very old, I followed it back to https://github.com/w3c/testharness.js/commit/1780607fc4582b65b7fcdded3e39e79e95c8d915 in 2012 but that was a copy.

Replacing it with `Object.is` would be cool, but not necessary as part of this.

Reply:
For historical interest, despite the fact that I'd forgotten `same_value` existed in the comment above, I'm pretty sure it was code I wrote at Opera, informed by the experience of working on Carakan (a rewrite of the js engine), where the edge cases it handles are very important to get right. So I predict it was in the earliest versions of `testharness.js` when ES3 was still the contemporaneous version.